### PR TITLE
add update from the polonius wg

### PR DESCRIPTION
**Pull request by lqd**
This post is an Inside Rust update from the polonius working group describing our roadmap and milestones, and recent progress. This was co-written with @nikomatsakis.

Feedback is very much appreciated.

### CODE REVIEW

File: posts/inside-rust/2023-09-29-polonius-update.md
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:12]:
```suggestion
Polonius refers to a few things. It is a [new formulation](http://smallcultfollowing.com/babysteps/blog/2018/04/27/an-alias-based-formulation-of-the-borrow-checker/) of the borrow checker. It is also a [specific project](https://github.com/rust-lang/polonius) that implemented that analysis, based on datalog. Our current plan does not make use of that datalog-based implementation, but uses what we learned implementing it to focus on reimplementing Polonius within rustc.
```

File: posts/inside-rust/2023-09-29-polonius-update.md
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:57]:
```suggestion
Issues like "NLL problem case #3", issue #47680 and others, were therefore deferred from NLLs, and left as future work, [Polonius](http://smallcultfollowing.com/babysteps/blog/2018/04/27/an-alias-based-formulation-of-the-borrow-checker/).
```

File: posts/inside-rust/2023-09-29-polonius-update.md
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:78]:
```suggestion
In the shorter term, we can refactor the borrow checker to separate out the higher-ranked processing from the ordinary processing. The goal would be to preprocess the outlives constraints in a kind of Polonius ["leak check"](https://rustc-dev-guide.rust-lang.org/traits/hrtb.html#basic-matching-and-placeholder-leaks), where we can compute the higher-ranked errors. This can then be removed once the trait solver can solve these constraints.
```

File: posts/inside-rust/2023-09-29-polonius-update.md
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:85]:
```suggestion
Out of the two key differences between polonius and the existing borrow check (regions as "sets of loans", and computing subtyping relations at each point in the CFG), this step is aimed at resolving the *first* difference, but not the second, so we call it the "location *in*sensitive loans in scope" (because subtyping is being done once, not per location): the idea can be described as "NLLs with the Polonius model".
```

File: posts/inside-rust/2023-09-29-polonius-update.md
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:101]: The wording here is a bit weird

File: posts/inside-rust/2023-09-29-polonius-update.md
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:127]:
```suggestion
The Types team is building a model of Rust's MIR and trait system called [`a-mir-formality`][]. Once it reaches a sufficiently complete status, the intent is that the model will always be extended to cover new language features prior to stabilization. We are therefore working to add Polonius into the model. This will in fact be the second time doing such modeling, as we already added Polonius to a previous incarnation of a-mir-formality. In fact, that modeling work is what gave us the insights that enabled the location-insensitive polonius formulation now landing on nightly.
```

File: posts/inside-rust/2023-09-29-polonius-update.md
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:141]:
```suggestion
Around this time, librarification efforts can also be rebooted, to turn the in-tree Polonius into a library, maybe using [Stable MIR][]. This is so that it could be reused elsewhere, for example in [rust-analyzer][], or [gccrs][], or by researchers working on verification tools (like [kani][]), [prusti][] and [creusot][]).
```

Comment (jackh726) [reply posts/inside-rust/2023-09-29-polonius-update.md:141]: Also, missing links here.

File: posts/inside-rust/2023-09-29-polonius-update.md
Comment (jackh726) [inline posts/inside-rust/2023-09-29-polonius-update.md:152]:
```suggestion
We are very excited to see the plan for Polonius coming into focus. At the moment, as we are still doing foundational work, we are not looking for volunteers or contributors unless they are well versed in the compiler. We do expect that as the project proceeds, there will be more and more need for new contributions. Stay tuned for updates!
```

**Review by jackh726 — COMMENTED** [review]
Probably better to capitalize Polonius consistently, but not a huge deal I guess if its consistent.

**Comment by lqd** [comment]
We can consistently capitalize it here if we want, sure. I've pushed a commit to address this. (We often don't in practice, e.g. see Niko's [last post](https://smallcultfollowing.com/babysteps/blog/2023/09/22/polonius-part-1/)).

File: posts/inside-rust/2023-09-29-polonius-update.md
Comment (lqd) [reply posts/inside-rust/2023-09-29-polonius-update.md]: I've pushed a commit to address this.

File: posts/inside-rust/2023-09-29-polonius-update.md
Comment (lqd) [reply posts/inside-rust/2023-09-29-polonius-update.md]: I've pushed a commit to reword this into the following, which should be clearer:

> Current status: ⏳ in-progress, the crater run itself will be done just before the PR lands.

File: posts/inside-rust/2023-09-29-polonius-update.md
Comment (lqd) [reply posts/inside-rust/2023-09-29-polonius-update.md]: I've pushed a commit for the capitalization. The links are just below the sentence though.

**Review by jackh726 — APPROVED** [review]

File: posts/inside-rust/2023-09-29-polonius-update.md
Comment (jackh726) [reply posts/inside-rust/2023-09-29-polonius-update.md]: I probably would have removed the "just" here, but this is fine.

**Comment by jackh726** [comment]
@lqd did you want to wait until the 29th to land this? Given this is just inside rust, I'm happy to land whenever.

File: posts/inside-rust/2023-09-29-polonius-update.md
Comment (lqd) [reply posts/inside-rust/2023-09-29-polonius-update.md]: I can do that.

**Comment by lqd** [comment]
@jackh726 Not particularly, I only picked "a week later" to allow time for review. Happy to change it to whichever date is preferred.

**Comment by jackh726** [comment]
Let's aim for tomorrow, Monday!

**Comment by lqd** [comment]
Sounds good. I pushed a commit to update the publication date.

File: posts/inside-rust/2023-09-25-polonius-update.md
Comment (shepmaster) [inline posts/inside-rust/2023-09-25-polonius-update.md:141]:
```suggestion
Around this time, librarification efforts can also be rebooted, to turn the in-tree Polonius into a library, maybe using [Stable MIR][]. This is so that it could be reused elsewhere, for example in [rust-analyzer][], or [gccrs][], or by researchers working on verification tools (like [kani][], [prusti][] and [creusot][]).
```

Comment (lqd) [reply posts/inside-rust/2023-09-25-polonius-update.md]: Good catch, thank you.

File: posts/inside-rust/2023-09-25-polonius-update.md
Comment (calebcartwright) [inline posts/inside-rust/2023-09-25-polonius-update.md:72]: This isn't rendering for me on GitHub. Unsure if this is anticipated (e.g. just a quirk of paths that work on the site but not inline) but wanted to note in case it was typo

Comment (lqd) [reply posts/inside-rust/2023-09-25-polonius-update.md:72]: I don't believe images can render on github markdown previews because they are collected into a different web root `/static/` and this link pattern is used for resources in Inside Rust. 

Look at e.g. https://github.com/rust-lang/blog.rust-lang.org/pull/1103/files and https://github.com/rust-lang/blog.rust-lang.org/blob/92ac9e0679d4f41b338851aab1e18b33071bff9b/posts/inside-rust/2023-05-09-api-token-scopes.md?plain=1#L18 

FWIW the image does show up when locally serving the blog.

File: posts/inside-rust/2023-09-25-polonius-update.md
Comment (lcnr) [inline posts/inside-rust/2023-09-25-polonius-update.md:55]: how is the loan used in the `None` path? :think

File: posts/inside-rust/2023-09-25-polonius-update.md
Comment (lcnr) [inline posts/inside-rust/2023-09-25-polonius-update.md:78]: tbh I am a bit worried that this short term idea may not work out, idk if that means I would like to slightly reword this or just wait until we chat about it (hopefully today) :shrug:

Comment (nikomatsakis) [reply posts/inside-rust/2023-09-25-polonius-update.md:78]: Personally, I'd prefer not to wait to discuss -- I think that's putting too much weight into a status update blog post. But we could reword this to like "we are exploring refactoring the borrow checker" or something like that -- and "this could then be removed".

Comment (lcnr) [reply posts/inside-rust/2023-09-25-polonius-update.md:78]: yeah, changing it to "we're looking into refactoring" is also fine :+1: I mostly want to avoid definitive statements while not unsure whether this step is actually feasible

File: posts/inside-rust/2023-09-25-polonius-update.md
Comment (lqd) [reply posts/inside-rust/2023-09-25-polonius-update.md:55]: yeah good spot, this actually describes a different minimization of the same issue 

```rust
struct Foo;

impl Foo {
    fn get_self(&mut self) -> Option<&mut Self> {
        Some(self)
    }

    fn bad_method(&mut self) {
        let mut var = self;
        
        loop {
            var = match var.get_self() {
                Some(v) => v,
                None => var
            };
        }
    }
}

fn main() {}
```

It's not the simplest/clearest of the 5 minimizations though, so I'll rewrite the description instead of changing the code example. I likely won't have time to do so until Wednesday/Thursday however.

File: posts/inside-rust/2023-09-25-polonius-update.md
Comment (nikomatsakis) [reply posts/inside-rust/2023-09-25-polonius-update.md:78]:
```suggestion
In the shorter term, we are exploring refactoring the borrow checker to separate out the higher-ranked processing from the ordinary processing. The goal would be to preprocess the outlives constraints in a kind of Polonius ["leak check"](https://rustc-dev-guide.rust-lang.org/traits/hrtb.html#basic-matching-and-placeholder-leaks), where we can compute the higher-ranked errors. This could then be removed once the trait solver can solve these constraints.
```

File: posts/inside-rust/2023-09-25-polonius-update.md
Comment (lqd) [reply posts/inside-rust/2023-09-25-polonius-update.md]: I tried to describe the cause succinctly, but it ends up less clear and less accurate as a simple soundbite. Niko rewrote it into the few sentences I've pushed in https://github.com/rust-lang/blog.rust-lang.org/pull/1147/commits/ce90cd698a56441486b8cd572ea8550454cd4560. 

It's clearer now, albeit wordy -- which could be fine on Inside Rust, or better suited to a footnote.

@lcnr what do you think?

File: posts/inside-rust/2023-09-25-polonius-update.md
Comment (lqd) [reply posts/inside-rust/2023-09-25-polonius-update.md:80]: I can add that we can find additional details on terminology, on the problem itself, and on this future possible solution involving the trait solver, in this blog spot https://smallcultfollowing.com/babysteps/blog/2019/01/21/hereditary-harrop-region-constraints/.

**Review by lcnr — APPROVED** [review]

**Comment by lqd** [comment]
Now that lcnr's comment is resolved, I updated the publication date and squashed everything.

**Pull request merged**
